prov/psm2/configure.m4

```diff
@@ -74,3 +75,29 @@ AC_ARG_WITH([psm2-src],
                            [Provide path to the source code of PSM2 library
 			    to be compiled into the provider]))
 
+dnl Extract struct definition from header file
+dnl
+dnl $1: struct name
+dnl $2: header file
+dnl $3: output file
+dnl
+AC_DEFUN([FI_PSM2_GEN_STRUCT_DEF],[
+	count=`grep -A 1000 '^struct $1 {' $2 | grep -n -m 1 '^};' | grep -o '^[[[0-9]]]*'`
+	count=`expr $count - 1`
+	grep -A $count "^struct $1 {" $2 >>$3
+])
+
+dnl Generate definition of "struct psm2_mq_req" from PSM2 source
+dnl
+AC_DEFUN([FI_PSM2_GEN_INTERNAL_H],[
+	infile1=$srcdir/prov/psm2/src/psm2/psm_mq_internal.h
+	infile2=$srcdir/prov/psm2/src/psm2/psm_ep.h
+	outfile=$srcdir/prov/psm2/src/psm2/psm2_internal.h
+	$as_echo '#include "opa_queue.h"' >$outfile
+	grep '^#define NUM_HASH_CONFIGS' $infile1 >>$outfile
+	grep '^#define NUM_MQ_SUBLISTS' $infile1 >>$outfile
+	grep '^typedef.*callback_fn_t.*psm2_mq_req_t' $infile1 >>$outfile
+	FI_PSM2_GEN_STRUCT_DEF(mqq, $infile2, $outfile)
+	FI_PSM2_GEN_STRUCT_DEF(psm2_mq_req, $infile1, $outfile)
+])
```

Comment:
gads - this seems ugly -- can we just copy-paste the structures to an internal header file or something?

Reply:
It was originally done that way. However, being internal means that the structure may change and generating the definition from the source is more future proof.